### PR TITLE
Add beta_cdf Presto function 

### PR DESCRIPTION
**Pull request by Ann-Rose**
Resolves https://github.com/facebookincubator/velox/issues/4637

### CODE REVIEW

File: velox/functions/prestosql/tests/ArithmeticTest.cpp [hunk 1]
```diff
@@ -699,5 +699,26 @@ TEST_F(ArithmeticTest, truncate) {
   EXPECT_DOUBLE_EQ(truncate(123456789012345678901.23, -21).value(), 0.0);
 }
 
+TEST_F(ArithmeticTest, BetaCDF) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/ArithmeticTest.cpp:702]: Nit : test name "BetaCDF" should start with a lowercase letter.

File: velox/functions/prestosql/tests/ArithmeticTest.cpp [hunk 2]
```diff
@@ -699,5 +699,26 @@ TEST_F(ArithmeticTest, truncate) {
...
+  const auto beta_cdf = [&](std::optional<double> a,
+                            std::optional<double> b,
+                            std::optional<double> value) {
+    return evaluateOnce<double>("beta_cdf(c0, c1, c2)", a, b, value);
+  };
+
+  EXPECT_EQ(0.09888000000000001, beta_cdf(3, 4, 0.2));
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/ArithmeticTest.cpp:709]: Please add more test cases especially around boundary values, and special values like std::numeric_limits<T>::infinity(),
std::numeric_limits<T>::max()/min(), std::numeric_limits<T>:quiet_NaN()

**Review by aditi-pandit — COMMENTED** [review]
Thanks for adding this function !

File: velox/functions/prestosql/tests/ArithmeticTest.cpp (hunk 1 above)
Comment (Ann-Rose) [reply velox/functions/prestosql/tests/ArithmeticTest.cpp]: Fixed it.

File: velox/functions/prestosql/tests/ArithmeticTest.cpp (hunk 2 above)
Comment (Ann-Rose) [reply velox/functions/prestosql/tests/ArithmeticTest.cpp]: Added testcases.

**Comment by Ann-Rose** [comment]
@aditi-pandit Here is the output of the Fuzzer test.

```
./velox_expression_fuzzer_test  --seed 123456 --duration_sec 60 --logtostderr=1 --minloglevel=0  --only beta_cdf  --v=1 2>&1 | tee fuzzertest.out
...
I20230424 14:38:49.292706 1852340 ExpressionVerifier.cpp:57] ===================
I20230424 14:38:49.292723 1852340 ExpressionVerifier.cpp:68] All results match.
I20230424 14:38:49.292763 1852340 ExpressionFuzzer.cpp:1172] ==============================> Done with iteration 457
I20230424 14:38:49.292971 1852340 ExpressionFuzzer.cpp:972] ==============================> Top 1 by number of rows processed
I20230424 14:38:49.292984 1852340 ExpressionFuzzer.cpp:974] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230424 14:38:49.292990 1852340 ExpressionFuzzer.cpp:978] beta_cdf 47578 100.00% 11642
I20230424 14:38:49.293000 1852340 ExpressionFuzzer.cpp:984] ==============================> Bottom 1 by number of rows processed
I20230424 14:38:49.293007 1852340 ExpressionFuzzer.cpp:986] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230424 14:38:49.293015 1852340 ExpressionFuzzer.cpp:991] beta_cdf 47578 100.00% 11642
I20230424 14:38:49.293022 1852340 ExpressionFuzzer.cpp:1004] ==============================> All stats sorted by number of times the function was chosen
I20230424 14:38:49.293030 1852340 ExpressionFuzzer.cpp:1006] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230424 14:38:49.293035 1852340 ExpressionFuzzer.cpp:1010] beta_cdf 47578 100.00% 11642
```

**Comment by Ann-Rose** [comment]
@aditi-pandit I have made the changes. Can you please review?

File: velox/docs/functions/presto/math.rst [hunk 3]
```diff
@@ -6,6 +6,12 @@ Mathematical Functions
 
     Returns the absolute value of ``x``.
 
+.. function:: beta_cdf(a, b, value) -> double
```
Comment (mbasmanova) [inline velox/docs/functions/presto/math.rst:9]: Let's move this function to a new section "Probability Functions: cdf" like in https://prestodb.io/docs/current/functions/math.html

Let's also convert "Beta cdf" into a link probably to https://en.wikipedia.org/wiki/Beta_distribution

File: velox/functions/prestosql/Arithmetic.h [hunk 4]
```diff
@@ -23,15 +23,19 @@
 #include <functional>
 #include <system_error>
 
+#include "boost/math/distributions/beta.hpp"
 #include "folly/CPortability.h"
 #include "velox/common/base/Exceptions.h"
 #include "velox/functions/Macros.h"
 #include "velox/functions/prestosql/ArithmeticImpl.h"
 
+using boost::math::beta_distribution;
```
Comment (mbasmanova) [inline velox/functions/prestosql/Arithmetic.h:32]: Do not use "using" in header files.

File: velox/functions/prestosql/Arithmetic.h [hunk 5]
```diff
@@ -495,5 +499,23 @@ struct TruncateFunction {
   }
 };
 
+template <typename T>
+struct BetaCDFFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void
+  call(double& result, double a, double b, double value) {
+    VELOX_USER_CHECK((a > 0) && (b > 0), "alpha and beta must be > 0");
```
Comment (mbasmanova) [inline velox/functions/prestosql/Arithmetic.h:508]: Let's split this check into two separate checks and use VELOX_USER_CHECK_GT for better error messages.

Let's also match Presto's error messages:

```
presto:di> select beta_cdf(-1, -1, 0);
Query 20230426_132309_08035_wz4f2 failed: a must be > 0

presto:di> select beta_cdf(0.5, -1, 0);
Query 20230426_132317_08038_wz4f2 failed: b must be > 0
```

File: velox/functions/prestosql/Arithmetic.h [hunk 6]
```diff
@@ -495,5 +499,23 @@ struct TruncateFunction {
...
+    VELOX_USER_CHECK(
+        (value >= 0) && (value <= 1), "value must be in the interval [0, 1]");
+    VELOX_USER_CHECK(
+        (a != kInf) && (b != kInf),
+        "alpha, beta values can't accept infinity value");
```
Comment (mbasmanova) [inline velox/functions/prestosql/Arithmetic.h:513]: Presto seems to allow that:

```
presto:di> select beta_cdf(0.5, infinity(), 0);
 _col0
-------
   0.0
(1 row)
```

File: velox/functions/prestosql/Arithmetic.h [hunk 7]
```diff
@@ -23,15 +23,19 @@
 #include <functional>
 #include <system_error>
 
+#include "boost/math/distributions/beta.hpp"
```
Comment (mbasmanova) [inline velox/functions/prestosql/Arithmetic.h:26]: Perhaps, start a new header file for probably functions: Probability.h

File: velox/functions/prestosql/tests/ArithmeticTest.cpp [hunk 8]
```diff
@@ -699,5 +699,32 @@ TEST_F(ArithmeticTest, truncate) {
   EXPECT_DOUBLE_EQ(truncate(123456789012345678901.23, -21).value(), 0.0);
 }
 
+TEST_F(ArithmeticTest, betaCDF) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArithmeticTest.cpp:702]: Perhaps, create a new test for probability functions: ProbabilityTest.cpp

**Review by mbasmanova — COMMENTED** [review]
@Ann-Rose Thank you for the contribution. Some initial comments.

File: velox/functions/prestosql/tests/ArithmeticTest.cpp [hunk 9]
```diff
@@ -699,5 +699,32 @@ TEST_F(ArithmeticTest, truncate) {
...
+  const auto betaCDF = [&](std::optional<double> a,
+                           std::optional<double> b,
+                           std::optional<double> value) {
+    return evaluateOnce<double>("beta_cdf(c0, c1, c2)", a, b, value);
+  };
+
+  EXPECT_EQ(0.09888000000000001, betaCDF(3, 4, 0.2));
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/ArithmeticTest.cpp:709]: Please also add tests for
std::numeric_limits::infinity(), std::numeric_limits::max(), std::numeric_limits::min(), std::numeric_limits::quiet_NaN() for the different parameters.

File: velox/functions/prestosql/tests/ArithmeticTest.cpp (hunk 8 above)
Comment (Ann-Rose) [reply velox/functions/prestosql/tests/ArithmeticTest.cpp]: Created new testfile

File: velox/functions/prestosql/Arithmetic.h (hunk 7 above)
Comment (Ann-Rose) [reply velox/functions/prestosql/Arithmetic.h]: Created new file: Probability.h

File: velox/functions/prestosql/Arithmetic.h (hunk 4 above)
Comment (Ann-Rose) [reply velox/functions/prestosql/Arithmetic.h]: removed it

File: velox/docs/functions/presto/math.rst (hunk 3 above)
Comment (Ann-Rose) [reply velox/docs/functions/presto/math.rst]: moved beta_cdf into new section

File: velox/functions/prestosql/Arithmetic.h (hunk 5 above)
Comment (Ann-Rose) [reply velox/functions/prestosql/Arithmetic.h]: Fixed it

File: velox/functions/prestosql/Arithmetic.h (hunk 6 above)
Comment (Ann-Rose) [reply velox/functions/prestosql/Arithmetic.h]: Using boost library for this, which doesn't support infinity value for beta_cdf. I have added an if condition to add this.

File: velox/functions/prestosql/tests/ArithmeticTest.cpp (hunk 9 above)
Comment (Ann-Rose) [reply velox/functions/prestosql/tests/ArithmeticTest.cpp]: Done

**Comment by Ann-Rose** [comment]
@mbasmanova I have made the changes mentioned. Can you please review?

**Comment by mbasmanova** [comment]
@Ann-Rose I'm seeing CI failures. Would you rebase?

File: velox/docs/functions/presto/math.rst [hunk 10]
```diff
@@ -233,3 +233,14 @@ Floating Point Functions
...
     Returns the constant representing not-a-number.
+
+
+====================================
+Probability Functions
+====================================
+
+.. function:: beta_cdf(a, b, value) -> double
+
+    Compute the Beta cdf with given a, b parameters:  P(N < value; a, b).
```
Comment (mbasmanova) [inline velox/docs/functions/presto/math.rst:244]: Let's turn "Beta cdf" into a link to https://en.wikipedia.org/wiki/Beta_distribution

File: velox/functions/prestosql/Probability.h [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "boost/math/distributions/beta.hpp"
+
+namespace facebook::velox::functions {
+
+constexpr double kInf = std::numeric_limits<double>::infinity();
```
Comment (mbasmanova) [inline velox/functions/prestosql/Probability.h:22]: Move this inside 'call' method.

File: velox/functions/prestosql/tests/ProbabilityTest.cpp [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+  EXPECT_EQ(1.0, betaCDF(3, 3.6, 1.0));
+  EXPECT_EQ(0.21764809997679951, betaCDF(3, 3.6, 0.3));
+  EXPECT_EQ(0.9972502881611551, betaCDF(3, 3.6, 0.9));
+  EXPECT_EQ(0.0, betaCDF(kInf, 3, 0.2));
+  EXPECT_EQ(0.0, betaCDF(3, kInf, 0.2));
+  EXPECT_EQ(0.0, betaCDF(kInf, kInf, 0.2));
+  EXPECT_EQ(
+      0.0,
+      betaCDF(
+          std::numeric_limits<double>::max(),
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ProbabilityTest.cpp:56]: nit: add kMin and kMax constants for readability

**Review by mbasmanova — APPROVED** [review]
@Ann-Rose Looks great % a couple of nits.

**Comment by Ann-Rose** [comment]
> @Ann-Rose I'm seeing CI failures. Would you rebase?

Rebased it

File: velox/functions/prestosql/tests/ProbabilityTest.cpp (hunk 12 above)
Comment (Ann-Rose) [reply velox/functions/prestosql/tests/ProbabilityTest.cpp]: Fixed it

File: velox/functions/prestosql/Probability.h (hunk 11 above)
Comment (Ann-Rose) [reply velox/functions/prestosql/Probability.h]: @mbasmanova  Fixed it

File: velox/docs/functions/presto/math.rst (hunk 10 above)
Comment (Ann-Rose) [reply velox/docs/functions/presto/math.rst]: @mbasmanova Fixed it

**Comment by Ann-Rose** [comment]
@mbasmanova I have made all the changes. Thank you

**Comment by Yuhta** [comment]
Have we run fuzzer over this new function?

**Comment by Ann-Rose** [comment]
> Have we run fuzzer over this new function?

Output of the fuzzer test
```
I20230503 20:42:56.177747 506972 ExpressionVerifier.cpp:57] ===================
I20230503 20:42:56.177757 506972 ExpressionVerifier.cpp:68] All results match.
I20230503 20:42:56.177778 506972 ExpressionFuzzer.cpp:1275] ==============================> Done with iteration 466
I20230503 20:42:56.177929 506972 ExpressionFuzzer.cpp:1068] ==============================> Top 1 by number of rows processed
I20230503 20:42:56.177937 506972 ExpressionFuzzer.cpp:1070] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230503 20:42:56.177942 506972 ExpressionFuzzer.cpp:1074] beta_cdf 49148 100.00% 11642
I20230503 20:42:56.177951 506972 ExpressionFuzzer.cpp:1080] ==============================> Bottom 1 by number of rows processed
I20230503 20:42:56.177956 506972 ExpressionFuzzer.cpp:1082] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230503 20:42:56.177961 506972 ExpressionFuzzer.cpp:1087] beta_cdf 49148 100.00% 11642
I20230503 20:42:56.177968 506972 ExpressionFuzzer.cpp:1100] ==============================> All stats sorted by number of times the function was chosen
I20230503 20:42:56.177973 506972 ExpressionFuzzer.cpp:1102] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230503 20:42:56.177978 506972 ExpressionFuzzer.cpp:1106] beta_cdf 49148 100.00% 11642
```

File: velox/functions/prestosql/tests/ProbabilityTest.cpp [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+#include "velox/common/base/tests/GTestUtils.h"
+#include "velox/functions/prestosql/tests/utils/FunctionBaseTest.h"
+
+namespace facebook::velox {
+namespace {
+
+constexpr double kInf = std::numeric_limits<double>::infinity();
+constexpr double kNan = std::numeric_limits<double>::quiet_NaN();
+constexpr double kDoubleMax = std::numeric_limits<double>::max();
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/ProbabilityTest.cpp:27]: Rename : kMax, kMin

File: velox/functions/prestosql/tests/ProbabilityTest.cpp [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+class ProbabilityTest : public functions::test::FunctionBaseTest {};
+
+TEST_F(ProbabilityTest, betaCDF) {
+  const auto betaCDF = [&](std::optional<double> a,
+                           std::optional<double> b,
+                           std::optional<double> value) {
+    return evaluateOnce<double>("beta_cdf(c0, c1, c2)", a, b, value);
+  };
+
+  EXPECT_EQ(0.09888000000000001, betaCDF(3, 4, 0.2));
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/ProbabilityTest.cpp:47]: How did you compute the result values ?

File: velox/functions/prestosql/tests/ProbabilityTest.cpp [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+namespace facebook::velox {
+namespace {
+
+constexpr double kInf = std::numeric_limits<double>::infinity();
+constexpr double kNan = std::numeric_limits<double>::quiet_NaN();
+constexpr double kDoubleMax = std::numeric_limits<double>::max();
+constexpr double kDoubleMin = std::numeric_limits<double>::min();
+
+MATCHER(IsNan, "is NaN") {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/ProbabilityTest.cpp:30]: Why are these MATCHERS needed ?

**Pull request closed without merging**